### ISSUE TITLE
Json Export

### ISSUE THREAD

**Issue by ds5678** [AssetRipper/AssetRipper#861]
## Context

It would be desirable to support exporting assets as Json.
* Json is easier than Yaml to use outside of Unity.
* Yaml export could be refactored to use the Json type system and be removed as a dependency from the source generation.
* This might make it easier to resolve #821.

## Problems

* We want to support retention of important information like where to use flow mapping.
* Unity allows nonunique keys in dictionaries.

## Design: `@style`

Mapping style will be stored as a property on `JsonObject`. Options will be:
* `Block` (for both maps and sequences)
* `BlockCurve` (for sequences only)
* `Flow` (for both maps and sequences)
* `Raw` (for sequences only)

Flow mapping example:
```json
{
    "@style" : "Flow",
    "x" : 2.5,
    "y" : 3.14
}
```

Currently, AssetRipper only uses `BlockCurve` sequences for nontraditional mappings. Normal sequences use the `Block` style. However, it may be desirable to future proof and encode this extra information.

```json
{
    "@style" : "BlockCurve",
    "@sequence" : [ 1, 2, 3 ]
}
```

## Design: `@sequence`

Sequences will be stored as a sequence property on `JsonObject`.

```json
{
    "@sequence" : [ 4, 5, 6 ]
}
```

## Design: `@map`

Mappings will be stored as a sequence property on `JsonObject`.

Even when keys are unique, they are still stored as a sequence of pairs:
```json
{
    "@map" : [
    {
        "@pair" : true,
        "Key" : "Eins",
        "Value" : 1
    },
    {
        "@pair" : true,
        "Key" : "Zwei",
        "Value" : 2
    }
    ]
}
```
Keys can be types other than `string` and `Utf8String`:
```json
{
    "@map" : [
    {
        "@pair" : true,
        "Key" : 
        {
            "@style" : "Flow",
            "x" : 2.5,
            "y" : 3.14
        },
        "Value" : ""
    }
    ]
}
```

## Design: `@pair`

Pairs will be stored as properties on `JsonObject`. The Boolean `@pair` property distinguishes this from normal objects.

```json
{
    "@pair" : true,
    "Key" : "Drei",
    "Value" : 3
}
```

**Comment by ds5678** [comment]
1e7c99235e139a11f9d8e022729b6bc94abbd05f added a framework for the Json serialization, along with a default implementation for the methods. It remains to choose the `SerializationOptions` and implement full source generation for the methods.

**Comment by ds5678** [comment]
Mostly implemented in e3a9f2ec8923262e04fd753f66a4c8ed0f050916. Still waiting on #504
